indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java

```diff
@@ -1014,6 +897,254 @@ static DimensionSchema createDimensionSchema(
+            if (index.getMetadata() == null) {
+              throw new RE(
+                  "Index metadata doesn't exist for segment [%s]. Try providing explicit rollup, "
+                  + "queryGranularity, dimensionsSpec, and metricsSpec.", dataSegment.getId()
+              );
+            }
+
+            processRollup(index);
+            processQueryGranularity(index);
+            processDimensionsSpec(index);
+            processMetricsSpec(index);
+          }
+        }
+      }
+    }
+
+    public Boolean getRollup()
+    {
+      if (!needRollup) {
+        throw new ISE("Not computing rollup");
+      }
+
+      return rollup;
+    }
+
+    public Granularity getQueryGranularity()
+    {
+      if (!needQueryGranularity) {
+        throw new ISE("Not computing queryGranularity");
+      }
+
+      return queryGranularity;
+    }
+
+    public DimensionsSpec getDimensionsSpec()
+    {
+      if (!needDimensionsSpec) {
+        throw new ISE("Not computing dimensionsSpec");
+      }
+
+      final BiMap<Integer, String> orderedDims = uniqueDims.inverse();
+      final List<DimensionSchema> dimensionSchemas =
+          IntStream.range(0, orderedDims.size())
+                   .mapToObj(i -> {
+                     final String dimName = orderedDims.get(i);
+                     return Preconditions.checkNotNull(
+                         dimensionSchemaMap.get(dimName),
+                         "Cannot find dimension[%s] from dimensionSchemaMap",
+                         dimName
+                     );
+                   })
+                   .collect(Collectors.toList());
+
+      return new DimensionsSpec(dimensionSchemas);
+    }
+
+    public AggregatorFactory[] getMetricsSpec()
+    {
+      if (!needMetricsSpec) {
+        throw new ISE("Not computing metricsSpec");
+      }
+
+      if (aggregatorFactoryLists.isEmpty()) {
+        return new AggregatorFactory[0];
+      }
+
+      final AggregatorFactory[] mergedAggregators = AggregatorFactory.mergeAggregators(
+          aggregatorFactoryLists.stream()
+                                .map(xs -> xs.toArray(new AggregatorFactory[0]))
+                                .collect(Collectors.toList())
+      );
+
+      if (mergedAggregators == null) {
+        throw new ISE(
+            "Failed to merge existing aggregators when generating metricsSpec; "
+            + "try providing explicit metricsSpec"
+        );
+      }
+
+      return mergedAggregators;
+    }
+
+    /**
+     * Sort segments in order, such that we look at later segments prior to earlier ones. Useful when analyzing
+     * dimensions, as it allows us to take the latest value we see, and therefore prefer types from more recent
+     * segments, if there was a change.
+     */
+    private void sortSegmentsList()
+    {
+      segments.sort(
+          (o1, o2) -> Comparators.intervalsByStartThenEnd().compare(o2.lhs.getInterval(), o1.lhs.getInterval())
```

Comment:
Would this work?
```suggestion
          Comparators.intervalsByStartThenEnd().reversed()
```

Reply:
Something still needs to extract the `interval`. I changed it to this to try to improve clarity:

```
      segments.sort(
          Comparator.comparing(
              o -> o.lhs.getInterval(),
              Comparators.intervalsByStartThenEnd().reversed()
          )
      );
```